src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1069,6 +1094,37 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
             }
 
             return true;
+
+            bool isAllowedInitOnlySet(BoundExpression receiver)
+            {
+                // ok: new C() { InitOnlyProperty = ... }
+                if (receiver is BoundObjectOrCollectionValuePlaceholder)
+                {
+                    return true;
+                }
+
+                // bad: other.InitOnlyProperty = ...
+                if (!(receiver is BoundThisReference || receiver is BoundBaseReference))
+                {
+                    return false;
+                }
+
+                var containingMember = ContainingMemberOrLambda;
+                if (!(containingMember is MethodSymbol method))
+                {
+                    return false;
+                }
+
+                Debug.Assert(!method.IsStatic);
```

Comment:
>Debug.Assert(!method.IsStatic); [](start = 16, length = 31)

I do not think this assert is appropriate, it is likely to fail in error scenarios today or in the future. I think implementation will be more robust if we do not make assumptions like that and instead check conditions for real.